Fix change tracking when fetching from collections or associations

chewi opened the pull request:
This should close #295.

chewi commented:
Just rebased this for your convenience. Please merge! :smiley_cat:

chewi commented:
Rebased again!

chewi commented:
Hmmm. This newest fix only works with Rails 4.2+. I therefore made it optional by checking for the presence of a method but I assumed that the specs would always use the latest version. I forgot that Travis also checks older ones.

edtjones commented:
hi @chewi  I've just got commit access on this repo and plan to get back on track with maintenance over time - will be in touch soon!

zacharywelch commented on lib/her/model/attributes.rb:
```diff
@@ -25,6 +25,7 @@ def initialize(attributes={})
 
         attributes = self.class.default_scope.apply_to(attributes)
         assign_attributes(attributes)
+        @changed_attributes = {} if persisted?
```
This should be covered in [#404](https://github.com/remiprev/her/pull/411/files#diff-7280373369c4cbefb4c1e7c1df8b8a36R51).

zacharywelch commented on lib/her/model/associations/has_many_association.rb:
```diff
@@ -85,7 +85,15 @@ def create(attributes = {})
+              entry.send("#{inverse_of}=", @parent)
+
+              # This fix requires ActiveModel 4.2+.
+              if entry.respond_to?(:clear_attribute_changes, true)
```
nice 👍 

Once we update `ActiveModel` we should probably look at `clear_changes_information` after a record is fetched or reloaded and `changes_applied` after persistence.

zacharywelch commented on spec/model/dirty_spec.rb:
```diff
@@ -5,19 +5,31 @@
         builder.adapter :test do |stub|
-          stub.get("/users/1") { [200, {}, { id: 1, fullname: "Lindsay Fünke" }.to_json] }
-          stub.get("/users/2") { [200, {}, { id: 2, fullname: "Maeby Fünke" }.to_json] }
+          stub.get("/users")   { [200, {}, [user1, user2].to_json] }
```
Can we stick w/ the current, explicit style to keep the specs consistent? Also think it helps to see the json w/ each stub 😺

chewi commented:
Sorry, this fell beneath the pile. I have rebased this, fixed the specs as requested, and changed the second commit to not require AR 4.2+. I believe assigning directly to the attributes hash should be safe here.

I see that #411 hasn't been merged yet. They deal with the same issue but I think this solution handles more cases.

chewi commented on lib/her/model/attributes.rb:
```diff
@@ -32,6 +32,7 @@ def initialize(attributes={})
         attributes = self.class.default_scope.apply_to(attributes)
         assign_attributes(attributes)
         yield self if block_given?
+        @changed_attributes = {} if persisted?
```
Well spotted. About to head out the door. I think this may miss some cases but I have an idea that I will try on Monday. Thanks!

chewi replied:
I've tried to make this work but regardless of which approach I use, the `id` parameter is ignored here. I haven't figured out why yet and I'm very busy this week but hopefully I'll get time on Friday.

chewi replied:
Found the reason. In `use_setter_methods`, both `id` and the actual primary key name are treated as reserved so assignment methods are not created for them when the record is instantiated. They end up in the `unset_attributes` hash in the `assign_attributes` and simply get merged into the `@attributes` hash. This means that no change tracking is applied. If you directly assign new values using setter methods afterwards then change tracking does work because `method_missing` is invoked, which creates attribute methods without checking a reserved list first.

If I change `use_setter_methods` to not treat these as reserved then the above works but it breaks the `handles new resource with custom primary key` test. I need to find a way for the `id` method to not be redefined without blocking the creation of an `id=` method.

chewi commented:
Okay @zacharywelch, I've made a change to how primary keys are handled that I believe makes things cleaner and deals with the issue we were seeing above. I've also changed the handling of `@changed_attributes` as you suggested.

chewi commented:
Hmm looks like this approach needs adjusting for Rails 3.2. It's also not quite right for subclasses either. Bear with me.

chewi commented:
I've put in a restriction to avoid the subclass and Rails 3.2 issues. Haven't had time to fully test this yet but I thought I'd push it up anyway because my break next week. Maybe I'll test it out at home if I get a chance.

chewi commented:
Okay, I've this against our own test suite and it all looks good.

zacharywelch commented:
Unfortunately had to remove the primary key changes before merging to master, but we solved the original issue of freshly-loaded resources being marked as dirty 🎩 💯

Still open to a solution of the primary key problem that doesn't involve raising an error if anyone's interested in putting together a PR. Glad we were able to identify some of the challenges.

Thanks!

zacharywelch commented:
I may have found a solution to the primary key problem @chewi. Working on a PR.

chewi commented:
Thanks. Do you mean the `cannot change primary key to :id if ancestors use other name` exception? Is there an issue with it? I thought this was extremely unlikely to be encountered in practise. I did find that it perhaps didn't matter if the attribute aliases were not undone, meaning the exception could potentially be avoided, but I took what I felt was the safer route.

zacharywelch commented:
Right, would rather find a solution that solves both use cases and remains backwards compatible. In the process of testing I uncovered another pre-existing issue related to `primary_key` so it just needs some  ❤️. Lemme type up an issue so we don't lose track of our findings.

Thanks!
